Capillary pressure property classes: Part 1, van Genuchten capillary pressure with/without regularization

wenqing opened the pull request:
As titled.
So far there three capillary pressure models are considered:
* van Genuchetch
* Brooks Corey (PR #2892)
* Liakopoulos (PR  #2892)

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.3.1)? yes in PR #2892.
2. [x] Tests covering your feature were added? yes, all added functions have unit test.
3. [x] Any new feature or behavior change was documented? yes

wenqing commented:
In can be reviewed once PR #2858 is merged.

endJunction commented on Documentation/ProjectFile/properties/property/CapillaryPressureVanGenuchten/c_CapillaryPressureVanGenuchten.md:
```diff
@@ -0,0 +1 @@
+\copydoc MaterialPropertyLib::SaturationVanGenuchten
```
should be pointing to capillary pressure.

wenqing replied:
They are same by the same \copdoc. Anyway replaced it with CapillaryPressureVanGenuchten.

endJunction replied:
Not yet everywhere.

wenqing replied:
Yes, only class documentation was changed because the class members of   `_residual_saturation`,  ` _maximuml_saturation` are not directly taken from the parameter input, which are set as `residual_liquid_saturation`, `residual_gas_saturation` for the consistency with `createSaturationVanGenuchten`.
In the recent changes, I kept the copying the documentations these two parameters from that for  `createSaturationVanGenuchten`, and made the  the documentations of other three parameters be copied from the class.

By the way, I suggest to replace `residual_gas_saturation` of capillary pressure classes and saturation classes with `maximum_liquid_saturation` in future. Because for these models, saturation means liquid saturation, we do not need highlight  `residual_gas_saturation` there.

endJunction replied:
The problem with this approach of referring to the same documentation from different classes is that it creates dependencies which are not there. If the SaturationVanGenuchten is renamed/moved what ever, how many places do you need to change?
Keep things simple. There is no need to create any dependencies in such simple implementations.

wenqing replied:
Documented the left two parameters directly (see 113fa68).

endJunction commented on Documentation/ProjectFile/properties/property/CapillaryPressureVanGenuchten/t_regularization.md:
```diff
@@ -0,0 +1,3 @@
+An optional input that indicates whether the capillary pressure is regularized
+ or not.
+Also see \copydoc  MaterialPropertyLib::RegularizedCapillaryVanGenuchten
```
copydoc is wrong here. Use \sa  http://www.doxygen.nl/manual/commands.html#cmdsa

wenqing replied:
Changed. Besides, RegularizedCapillaryVanGenuchten is changed to CapillaryPressureRegularizedVanGenuchten.

endJunction commented on MaterialLib/MPL/CreateProperty.cpp:
```diff
@@ -113,6 +113,11 @@ std::unique_ptr<MaterialPropertyLib::Property> createProperty(
 
+    if (boost::iequals(property_type, "CapillaryPressureVanGenuchten"))
+    {
+        return createSaturationVanGenuchten(config);
```
Doesn't match. CapillaryPressure property is created, but saturation is returned.

wenqing replied:
Made it be inherited from MPL::Property.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,147 @@
+    PropertyDataType value(VariableArray const& variable_array,
+                           ParameterLib::SpatialPosition const& pos,
+                           double const t,
+                           double const dt) const;
```
override is missing.

wenqing replied:
I see that you prefer to derive class from Property (in current PR, this one is inherited by RegularizedCapillaryVanGenuchten).

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,147 @@
+namespace MaterialPropertyLib
+{
+ /// \copydoc MaterialPropertyLib::SaturationVanGenuchten
+class CapillaryPressureVanGenuchten : public SaturationVanGenuchten
```
derive from property

endJunction replied:
class hierarchies represent a IsA relation. It would mean here, that "capillary pressure" is "saturation", which is not good at all.

wenqing replied:
Changed the hierarch.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,147 @@
+                             double const t, double const dt) const;
+
+protected:
+    double const _pc_max;  ///< Maximum capillary pressiure
```
spelling

wenqing replied:
Corrected.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,147 @@
+ *   Part I: formulation and properties of the mathematical model. Comput Geosci
+ *   2013;17(2):431-442.
+ */
+class RegularizedCapillaryVanGenuchten final
```
only one property per translation unit.

wenqing replied:
Changed accordingly.

endJunction replied:
No, not changed yet. Two properties in same translation unit.

wenqing replied:
Updated.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,147 @@
+        1e-5;  ///< parameter in regularized van Genuchten model
+
+    /// Regularized van Genuchten capillary pressure-saturation Model
+    double getPcBarvGSg(double const Sg) const;
```
Comment does not match the function. A comment should describe a what a function does.

wenqing replied:
Improved the comments.

endJunction commented on Tests/MaterialLib/TestCapillaryPressureVanGenuchten.cpp:
```diff
@@ -0,0 +1,234 @@
+        const double pc1 = capillary_pressure_property.template value<double>(
+            variable_array, pos, t, dt);
+        const double numerical_dPcdS = (pc1 - pc) / (S1 - S);
+        ASSERT_LE(std::fabs((dPcdS - numerical_dPcdS) / dPcdS), 1e-4)
```
1e-4 is rather crude estimate for a derivative.

wenqing replied:
The saturation is generated randomly. If saturation is close to its residual value, the capillary pressure and its derivative become very large (infinite). This gives the low accuracy of the numerical  derivative at S->S_r. If we fixed some saturation values for the unit test, the tolerance can be reduced to a very small value.

wenqing replied:
Besides, the value of dpc/dS or `d^2pc/dS^2` is large too. For example
```
Expected: (std::fabs((dPcdS - numerical_dPcdS) / dPcdS)) <= (1e-6), actual: 1.03874e-06 vs 1e-06
for the analytic derivative of dPc/dS -68935.282428887251 and numeric derivative of dPc/dS.-68935.354034423828
```

wenqing commented:
@endJunction  What is the following error that occurred during Pre-checks? I did not create a file with non-ASCII file name.
```
[2020-03-30T14:20:08.484Z] Check file names.........................................................Failed
[2020-03-30T14:20:08.484Z] hookid: file-names
[2020-03-30T14:20:08.484Z] 
[2020-03-30T14:20:08.484Z] scripts/hooks/pre-commit-file-names.sh: line 25: scripts/cmake/test/MoveDataToStore.cmake: Permission denied
[2020-03-30T14:20:08.484Z] Error: Attempt to add a non-ASCII file name.
```

endJunction commented:
@wenqing Sorry, I don't yet know how the error comes. It has to do with the file-names hook. Suspicion is that there is a file with a non-ascii char in the name. Don't know which one yet.

wenqing commented:
> @wenqing Sorry, I don't yet know how the error comes. It has to do with the file-names hook. Suspicion is that there is a file with a non-ascii char in the name. Don't know which one yet.

Thanks. The only new modified file to the latest changes in this PR is bibliography.bib. I am trying to figure out how this happens.

wenqing commented:
This file name did not pass the file name check: TestCapillaryPressureVanGenuchten.cpp. It is weird that it passed the check before the recent revision of this PR.

wenqing commented:
> @endJunction What is the following error that occurred during Pre-checks? I did not create a file with non-ASCII file name.
> 
> ```
> [2020-03-30T14:20:08.484Z] Check file names.........................................................Failed
> [2020-03-30T14:20:08.484Z] hookid: file-names
> [2020-03-30T14:20:08.484Z] 
> [2020-03-30T14:20:08.484Z] scripts/hooks/pre-commit-file-names.sh: line 25: scripts/cmake/test/MoveDataToStore.cmake: Permission denied
> [2020-03-30T14:20:08.484Z] Error: Attempt to add a non-ASCII file name.
> ```
@bilke Do you know the reason? I've tested by commits. Temporarily, I removed all commits after the failed commit, d73b00a. For all commit before commit d73b00a, Pre-checks works.  Commit d73b00a
only adds a file with the file name of TestCapillaryPressureVanGenuchten.cpp, which seems be blamed by Pre-checks.

wenqing commented:
Moved more commits to this PR:
* Brooks Corey
* Liakopoulos
Replaced  the file of `TestCapillaryPressureVanGenuchten.cpp` with `TestCapillaryPressure.cpp`, in which all three models are tested by using  `testing::Test`.

The error in Pre-checks is gone now. Maybe it is due to rebasing onto master not because of  removing  `TestCapillaryPressureVanGenuchten.cpp`.

wenqing commented:
Since any new TEST file (in Tests/MaterialLib) added at the end of this PR causes a file name problem in Pre-check, the ctest for this PR is moved to the next PR (#2892).

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,152 @@
+class Phase;
+class Component;
+
+/// \copydoc MaterialPropertyLib::SaturationVanGenuchten
+class CapillaryPressureVanGenuchten : public Property
```
These two do not fit.

wenqing replied:
CapillaryPressureVanGenuchten and SaturationVanGenuchten share the same documentation. Replace `\copydoc` with `\sa`?

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,152 @@
+                "The property 'CapillaryVanGenuchten' is implemented on the "
+                "'media' scale only.");
+        }
+        this->_medium = std::get<Medium*>(scale_pointer);
```
Is `this` needed?

wenqing replied:
removed.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,152 @@
+        this->_medium = std::get<Medium*>(scale_pointer);
+    }
+
+    /// gets \f$ p_c(S) \f$.
```
Returns/computes/calculates sounds much better to me, and actually reflect the true functions' purpose much better.

wenqing replied:
Changed to `It returns`.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,152 @@
+                "The property 'CapillaryVanGenuchten' is implemented on the "
+                "'media' scale only.");
+        }
+        this->_medium = std::get<Medium*>(scale_pointer);
```
Is `this` needed?

wenqing replied:
removed.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.h:
```diff
@@ -0,0 +1,152 @@
+    /// Residual saturation of liquid phase.
+    double const _residual_saturation;
+    double const _maximuml_saturation;  ///< Maximum saturation of liquid phase.
+    double const _m;                    ///< Exponent index.
```
Either exponent or index, but not both.

wenqing replied:
Changed.

wenqing commented on Documentation/ProjectFile/properties/property/CapillaryPressureVanGenuchten/t_entry_pressure.md:
```diff
@@ -0,0 +1 @@
+\copydoc ogs_file_param__properties__property__SaturationVanGenuchten__entry_pressure
```
Then I suggest two options:
1. use the original alpha, which has physical meaning. With that, you need to pass rho to your changes of value, dValue and d2Value  in PR #2880, and assign rho g/alpha to _p_b
2. use this present `_entry_pressure` and change it to `bulb pressure` or other meaning full name. 

We can also stick to 1. If there are some cases with p_b, we can convert it to alpha in advance.

For the current PR, the preset name and the use of p_b can be kept. The change can be  done once  PR #2880 is finished.

wenqing commented:
Replaced `entry_pressure` with `p_d` by amending the corresponding commits.

boyanmeng commented:
Very useful property. In some cases the capillary pressure is not used as a primary variable and therefore needs to be computed from saturation. Hope to see this PR get merged soon.

endJunction commented on MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.cpp:
```diff
@@ -0,0 +1,242 @@
+    : _residual_saturation(residual_liquid_saturation),
+      _maximuml_saturation(1. - residual_gas_saturation),
+      _m(exponent),
+      _p_b(entry_pressure),
```
not entry pressure...

wenqing replied:
Changed by amending commits.

wenqing commented:
New change:  [MPL/vanGenuchten] added checkVanGenuchtenExponentRange
014a6da

endJunction commented on Documentation/ProjectFile/properties/property/CapillaryPressureVanGenuchten/t_regularization.md:
```diff
@@ -0,0 +1,3 @@
+An optional input that indicates whether the capillary pressure is regularized
```
Just a possibility for now, don't know what others think. @Scinopode ?
What reads better:
```xml
<property>
    <type>CapillaryPressureVanGenuchtenRegularized</type>
    <!-- other settings -->
</property>
```
or 
```xml
<property>
    <type>CapillaryPressureVanGenuchten</type>
    <regularization>true</regularization>
    <!-- other settings -->
</property>
```
My preference is, of course, the extra property `CapillaryPressureVanGenuchtenRegularized`.

wenqing commented:
Mac-GUI file name error.
```
[Clang (macOS, GUI)] [-ERROR-] - '../MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.h', IO exception has been thrown:
```

endJunction commented:
@wenqing Maybe simply a time limit of 6h was hit...
There are warnings, though, on unused private members: https://jenkins.opengeosys.org/job/ufz/job/ogs/job/PR-2874/61/clang-mac-gui/

wenqing commented:
@endJunction  Changes:
1. Renamed `CapillaryPressureRegularizedVanGenuchten` to `RegularizedCapillaryPressureVanGenuchten` for naming issue (or better naming), and removed `_max_pc` from the class, 
c1b8b9e.
2. Added a creator for `RegularizedCapillaryPressureVanGenuchten`, 
90ec40b.
3. Replaced `residual_gas_saturation` with `maximum_liquid_saturation`.

endJunction commented:
Sorry, but previous naming was better, in my opinion, because consistent with currently implemented properties.
It is not the capillary pressure, which is regularized, but the van Genuchten model.
Residual gas saturation is consistent with currently available properties: https://github.com/ufz/ogs/blob/master/MaterialLib/MPL/PropertyType.h#L69-L70.

I'll fix the remaining issues myself and push it here.

wenqing commented:
> Sorry, but previous naming was better, in my opinion, because consistent with currently implemented properties.

The previous naming is better but I am not sure whether it cases "Mac-GUI file name error".  Changing it back is good if there is no naming checking issue on Mac-GUI.

> Residual gas saturation is consistent with currently available properties: https://github.com/ufz/ogs/blob/master/MaterialLib/MPL/PropertyType.h#L69-L70.

`enum PropertyType ` contains property types. However `residual_gas_saturation` or `maximum_liquid_saturation` for this property just a name of an input parameter. When  `residual_gas_saturation` is used, the value by it is immediately  converted to `maximum_liquid_saturation` in ctors of the classes of saturation properties or capillary pressure proprieties. Besides, the function is P(S_l), the input of S_l^r and S_l^{max} is more straightforward.
Therefore, I think `maximum_liquid_saturation`  is a better choice as a parameter of capillary pressure  models.

> I'll fix the remaining issues myself and push it here.
Thanks!

wenqing commented:
No naming  issue on Mac-GUI but 
[  FAILED  ] MaterialLibSolidsKelvinVector4.Inverse (irrelative to this PR).
>I also suggest to replace residual_gas_saturation with maximum_liquid_saturation in all classes of saturation classes. If you agree on this point, I can do it late.
Anyway if you and other guys think that residual_gas_saturation is better, it is OK to use it.

wenqing commented:
@endJunction  If you have not done any changes in this PR, I can push my changes of the name-back, which was just pushed to #2892 as a test for file naming. If you have done it, please just push yours.

wenqing commented:
@endJunction  I am go to push the changes of the file name-back  because that  Boyan wants to use these models in this PR urgently. For the parameter of `maximum_liquid_saturation` or `residual_gas_saturation`, I prefer to use `maximum_liquid_saturation` in this PR based the reasons that I addressed above. If you have other fixings, please push yours first. I hope it can be merged tomorrow.
